File: tools/hangar/src-tauri/src/troubleshoot.rs
use serde::Serialize;
use std::process::Command;

#[derive(Debug, Serialize, Clone)]
pub struct DetectedProcess {
    pub pid: u32,
    pub command: String,
}

#[derive(Debug, Serialize)]
pub struct KillOutcome {
    pub pid: u32,
    pub gone: bool,
    pub used_kill: bool,
    pub error: Option<String>,
}

fn run_capture(program: &str, args: &[&str]) -> Result<String, String> {
    let out = Command::new(program)
        .args(args)
        .output()
        .map_err(|e| format!("{program}: {e}"))?;
    if !out.status.success() {
        // lsof returns 1 with empty stdout when nothing matches —
        // that's not really an error for our use case, treat empty
        // as empty rather than bubbling up.
        if out.stdout.is_empty() && out.stderr.is_empty() {
            return Ok(String::new());
        }
        // pgrep exits 1 when no match. Same idea.
        if out.stdout.is_empty() {
            return Ok(String::new());
        }
    }
    Ok(String::from_utf8_lossy(&out.stdout).to_string())
}

fn pid_command(pid: u32) -> String {
    // ps -p <pid> -o command= prints just the command line (no header).
    // Use a sane fallback if the process has disappeared between when we
    // listed it and when we ask for its command.
    Command::new("ps")
        .args(["-p", &pid.to_string(), "-o", "command="])
        .output()
        .ok()
        .and_then(|o| {
            let s = String::from_utf8_lossy(&o.stdout).trim().to_string();
            if s.is_empty() { None } else { Some(s) }
        })
        .unwrap_or_else(|| "(process exited)".into())
}

#[tauri::command(rename_all = "snake_case")]
pub fn troubleshoot_scan_port(port: u16) -> Result<Vec<DetectedProcess>, String> {
    // -iTCP:<port> filters to the port (must be one combined arg — lsof
    // treats `:<port>` as a filename if it's a separate argv);
    // -t outputs just PIDs (one per line); -P avoids slow service-name
    // resolution; -n avoids slow DNS lookups.
    let i_filter = format!("-iTCP:{port}");
    let raw = run_capture(
        "lsof",
        &["-nP", &i_filter, "-sTCP:LISTEN", "-t"],
    )?;
    let mut out: Vec<DetectedProcess> = Vec::new();
    let mut seen: std::collections::HashSet<u32> = std::collections::HashSet::new();
    for line in raw.lines() {
        let pid: u32 = match line.trim().parse() {
            Ok(p) => p,
            Err(_) => continue,
        };
        if !seen.insert(pid) {
            continue;
        }
        out.push(DetectedProcess {
            pid,
            command: pid_command(pid),
        });
    }
    Ok(out)
}

#[tauri::command(rename_all = "snake_case")]
pub fn troubleshoot_scan_pattern(
    pattern: String,
) -> Result<Vec<DetectedProcess>, String> {
    // pgrep -f matches the full command line. -l would include the
    // process name; we fetch command via ps separately for consistency
    // with the port-based scan.
    let raw = run_capture("pgrep", &["-f", &pattern])?;
    let mut out: Vec<DetectedProcess> = Vec::new();
    let self_pid = std::process::id();
    for line in raw.lines() {
        let pid: u32 = match line.trim().parse() {
            Ok(p) => p,
            Err(_) => continue,
        };
        // Don't list ourselves — if the pattern is "fleet" we might
        // match Fleet Hangar's own command line.
        if pid == self_pid {
            continue;
        }
        out.push(DetectedProcess {
            pid,
            command: pid_command(pid),
        });
    }
    Ok(out)
}

#[cfg(unix)]
fn signal_pid(pid: u32, sig: libc::c_int) -> Result<(), String> {
    let r = unsafe { libc::kill(pid as libc::pid_t, sig) };
    if r != 0 {
        // ESRCH (3) means the process is already gone — not an error
        // for our caller; the post-signal poll handles the "did it
        // actually die" question.
        let err = std::io::Error::last_os_error();
        if err.raw_os_error() != Some(libc::ESRCH) {
            return Err(format!("kill({pid}, {sig}): {err}"));
        }
    }
    Ok(())
}

#[cfg(unix)]
fn pid_alive(pid: u32) -> bool {
    // Signal 0 doesn't deliver but checks permission/existence: 0 = exists,
    // ESRCH = gone, EPERM = exists but we can't signal it.
    let r = unsafe { libc::kill(pid as libc::pid_t, 0) };
    if r == 0 {
        return true;
    }
    let err = std::io::Error::last_os_error();
    err.raw_os_error() != Some(libc::ESRCH)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn troubleshoot_kill_pid(pid: u32) -> Result<KillOutcome, String> {
    #[cfg(not(unix))]
    {
        let _ = pid;
        return Err("kill is only implemented on unix platforms".into());
    }
    #[cfg(unix)]
    {
        if let Err(e) = signal_pid(pid, libc::SIGTERM) {
            return Ok(KillOutcome {
                pid,
                gone: !pid_alive(pid),
                used_kill: false,
                error: Some(e),
            });
        }
        // Poll for graceful exit. 2s total, 100ms granularity — enough
        // for well-behaved daemons (fleet serve, ngrok, python) without
        // making the user wait forever if SIGTERM is being ignored.
        for _ in 0..20 {
            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
            if !pid_alive(pid) {
                return Ok(KillOutcome {
                    pid,
                    gone: true,
                    used_kill: false,
                    error: None,
                });
            }
        }
        // Escalate.
        if let Err(e) = signal_pid(pid, libc::SIGKILL) {
            return Ok(KillOutcome {
                pid,
                gone: !pid_alive(pid),
                used_kill: true,
                error: Some(e),
            });
        }
        // Brief settle window after SIGKILL — the kernel cleans up
        // basically instantly, but we poll once more to be honest in
        // the response.
        for _ in 0..5 {
            tokio::time::sleep(std::time::Duration::from_millis(50)).await;
            if !pid_alive(pid) {
                return Ok(KillOutcome {
                    pid,
                    gone: true,
                    used_kill: true,
                    error: None,
                });
            }
        }
        Ok(KillOutcome {
            pid,
            gone: false,
            used_kill: true,
            error: Some("process still alive after SIGKILL".into()),
        })
    }
}

File: tools/hangar/internal/troubleshoot/troubleshoot.go
// Package troubleshoot scans for processes by listening port or command
// pattern and can terminate them. Ported from src-tauri/src/troubleshoot.rs.
// macOS-only (lsof/pgrep/ps + POSIX signals).
package troubleshoot

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"syscall"
	"time"
)

// DetectedProcess is a process found by a scan.
type DetectedProcess struct {
	PID     uint32 `json:"pid"`
	Command string `json:"command"`
}

// KillOutcome reports the result of terminating a pid.
type KillOutcome struct {
	PID      uint32  `json:"pid"`
	Gone     bool    `json:"gone"`
	UsedKill bool    `json:"used_kill"`
	Error    *string `json:"error"`
}

// runCapture runs program and returns stdout. lsof/pgrep exit 1 with empty
// output when nothing matches — that's "no results", not an error.
func runCapture(program string, args ...string) (string, error) {
	out, err := exec.Command(program, args...).Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			// Empty stdout on a non-zero exit = no matches.
			if len(out) == 0 && len(ee.Stderr) == 0 {
				return "", nil
			}
			if len(out) == 0 {
				return "", nil
			}
		} else {
			return "", fmt.Errorf("%s: %w", program, err)
		}
	}
	return string(out), nil
}

// parsePIDs extracts numeric PIDs from one-per-line output, skipping blanks
// and non-numeric lines.
func parsePIDs(raw string) []uint32 {
	var out []uint32
	for _, line := range strings.Split(raw, "\n") {
		n, err := strconv.ParseUint(strings.TrimSpace(line), 10, 32)
		if err != nil {
			continue
		}
		out = append(out, uint32(n))
	}
	return out
}

// detected turns PIDs into DetectedProcess entries. dedup drops repeats
// (keeping first); exclude (nonzero) skips a pid (e.g. our own). cmd
// resolves each pid's command line.
//
// PIDs that have already exited are skipped: pgrep/lsof can match a process
// that's mid-teardown (exactly the state right after stopping a server), and a
// dead process is neither killable nor a real orphan. This keeps the scan
// self-consistent — it reports only live matches, matching what a re-scan a
// moment later would show.
func detected(pids []uint32, dedup bool, exclude uint32, alive func(uint32) bool, cmd func(uint32) string) []DetectedProcess {
	var out []DetectedProcess
	seen := map[uint32]bool{}
	for _, pid := range pids {
		if exclude != 0 && pid == exclude {
			continue
		}
		if dedup {
			if seen[pid] {
				continue
			}
			seen[pid] = true
		}
		if !alive(pid) {
			continue
		}
		c := cmd(pid)
		if c == "(process exited)" {
			// Raced: exited between the liveness check and reading its command.
			continue
		}
		out = append(out, DetectedProcess{PID: pid, Command: c})
	}
	return out
}

// pidCommand returns the command line for pid via `ps`, or a placeholder if
// the process has exited.
func pidCommand(pid uint32) string {
	out, err := exec.Command("ps", "-p", strconv.FormatUint(uint64(pid), 10), "-o", "command=").Output()
	if err != nil {
		return "(process exited)"
	}
	s := strings.TrimSpace(string(out))
	if s == "" {
		return "(process exited)"
	}
	return s
}

// ScanPort lists processes listening on the given TCP port (deduped).
func ScanPort(port uint16) ([]DetectedProcess, error) {
	raw, err := runCapture("lsof", "-nP", fmt.Sprintf("-iTCP:%d", port), "-sTCP:LISTEN", "-t")
	if err != nil {
		return nil, err
	}
	return detected(parsePIDs(raw), true, 0, pidAlive, pidCommand), nil
}

// ScanPattern lists processes whose full command line matches pattern,
// excluding our own process (so searching "fleet" doesn't list Hangar).
func ScanPattern(pattern string) ([]DetectedProcess, error) {
	raw, err := runCapture("pgrep", "-f", pattern)
	if err != nil {
		return nil, err
	}
	return detected(parsePIDs(raw), false, uint32(os.Getpid()), pidAlive, pidCommand), nil
}

func signalPID(pid uint32, sig syscall.Signal) error {
	err := syscall.Kill(int(pid), sig)
	if err != nil && !errors.Is(err, syscall.ESRCH) {
		return fmt.Errorf("kill(%d, %d): %w", pid, sig, err)
	}
	return nil
}

// pidAlive reports whether pid exists. Signal 0 checks existence/permission
// without delivering anything: nil = exists, ESRCH = gone, EPERM = exists.
func pidAlive(pid uint32) bool {
	err := syscall.Kill(int(pid), 0)
	if err == nil {
		return true
	}
	return !errors.Is(err, syscall.ESRCH)
}

// KillPID sends SIGTERM, waits up to 2s for graceful exit, then escalates to
// SIGKILL.
func KillPID(pid uint32) KillOutcome {
	strp := func(s string) *string { return &s }

	if err := signalPID(pid, syscall.SIGTERM); err != nil {
		return KillOutcome{PID: pid, Gone: !pidAlive(pid), UsedKill: false, Error: strp(err.Error())}
	}
	for range 20 {
		time.Sleep(100 * time.Millisecond)
		if !pidAlive(pid) {
			return KillOutcome{PID: pid, Gone: true}
		}
	}
	if err := signalPID(pid, syscall.SIGKILL); err != nil {
		return KillOutcome{PID: pid, Gone: !pidAlive(pid), UsedKill: true, Error: strp(err.Error())}
	}
	for range 5 {
		time.Sleep(50 * time.Millisecond)
		if !pidAlive(pid) {
			return KillOutcome{PID: pid, Gone: true, UsedKill: true}
		}
	}
	return KillOutcome{PID: pid, Gone: false, UsedKill: true, Error: strp("process still alive after SIGKILL")}
}
